app/src/main/java/org/mozilla/fenix/components/history/PagedHistoryProvider.kt

```diff
@@ -129,9 +142,17 @@ class DefaultPagedHistoryProvider(
         }
         val historyMetadata = historyGroupsInOffset.flatMap { it.items }
 
-        // Add all items that are not in a group filtering out any matches with a history metadata
-        // item.
-        result.addAll(history.filter { item -> historyMetadata.find { it.url == item.url } == null })
+        // Add all history items that are not in a group filtering out any matches with a history
+        // metadata item.
+        for (historyItem in history) {
+            val matchedMetadata = historyMetadata.find { it.url == historyItem.url }
```

Comment:
This matching logic isn't sufficient yet. You could have `n` visits to the same page, and they are returned in desc order by visit timestamp. So you're basically always saving the oldest visit in the map currently which would then get deleted.

I think what we need is similar to the logic we already have, a mapping based on `visitedAt` and `createdAt` (for metadata). So for every metadata record we delete we will need to look for the corresponding visit and delete it too.